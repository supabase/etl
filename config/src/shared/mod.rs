mod base;
mod connection;
mod destination;
mod pipeline;
mod replicator;
mod supabase;

pub use base::*;
pub use connection::*;
pub use destination::*;
pub use pipeline::*;
pub use replicator::*;
pub use supabase::*;
