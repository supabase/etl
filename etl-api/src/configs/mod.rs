pub mod destination;
pub mod encryption;
pub mod pipeline;
pub mod serde;
pub mod source;
