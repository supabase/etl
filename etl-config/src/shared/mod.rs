mod base;
mod batch;
mod connection;
mod destination;
mod pipeline;
mod replicator;
mod retry;
mod sentry;
mod supabase;

pub use base::*;
pub use batch::*;
pub use connection::*;
pub use destination::*;
pub use pipeline::*;
pub use replicator::*;
pub use retry::*;
pub use sentry::*;
pub use supabase::*;
