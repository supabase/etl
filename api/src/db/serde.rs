use crate::encryption::{
    Decryptable, DecryptionError, Encryptable, EncryptionError, EncryptionKey,
};
use serde::de::DeserializeOwned;
use serde::Serialize;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum DbSerializationError {
    #[error("Error while serializing data to the db: {0}")]
    Serde(#[from] serde_json::Error),

    #[error("An error occurred while encrypting data for the db representation: {0}")]
    Encryption(#[from] EncryptionError),
}

#[derive(Debug, Error)]
pub enum DbDeserializationError {
    #[error("Error while deserializing data from the db: {0}")]
    Serde(#[from] serde_json::Error),

    #[error("An error occurred while decrypting data from the db representation: {0}")]
    Decryption(#[from] DecryptionError),
}

pub fn serialize<S>(value: S) -> Result<serde_json::Value, DbSerializationError>
where
    S: Serialize,
{
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

pub fn encrypt_and_serialize<T, S>(
    value: T,
    encryption_key: &EncryptionKey,
) -> Result<serde_json::Value, DbSerializationError>
where
    T: Encryptable<S>,
    S: Serialize,
{
    let value = value.encrypt(encryption_key)?;
    let serialized_value = serde_json::to_value(value)?;

    Ok(serialized_value)
}

pub fn deserialize_from_value<S>(value: serde_json::Value) -> Result<S, DbDeserializationError>
where
    S: DeserializeOwned,
{
    let deserialized_value = serde_json::from_value(value)?;

    Ok(deserialized_value)
}

pub fn decrypt_and_deserialize_from_value<T, S>(
    value: serde_json::Value,
    encryption_key: &EncryptionKey,
) -> Result<S, DbDeserializationError>
where
    T: Decryptable<S>,
    T: DeserializeOwned,
{
    let deserialized_value: T = serde_json::from_value(value)?;
    let value = deserialized_value.decrypt(encryption_key)?;

    Ok(value)
}
