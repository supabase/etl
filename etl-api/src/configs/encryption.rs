use std::string;

use aws_lc_rs::{
    aead::{AES_256_GCM, Aad, Nonce, RandomizedNonceKey},
    rand::fill,
};
use base64::{Engine, prelude::BASE64_STANDARD};
use serde::{Deserialize, Serialize};
use thiserror::Error;

/// Errors that occur during data encryption.
#[derive(Debug, Error)]
pub enum EncryptionError {
    /// An unspecified error occurred while encrypting data.
    #[error("An unspecified error occurred while encrypting data")]
    Unspecified(#[from] aws_lc_rs::error::Unspecified),
}

/// Errors that occur during data decryption.
#[derive(Debug, Error)]
pub enum DecryptionError {
    /// An unspecified error occurred while decrypting data.
    #[error("An unspecified error occurred while decrypting data")]
    Unspecified(#[from] aws_lc_rs::error::Unspecified),

    /// Failed to decode base64 data during decryption.
    #[error("An error occurred while decoding BASE64 data for decryption: {0}")]
    Decode(#[from] base64::DecodeError),

    /// Failed to convert decrypted bytes to UTF-8 string.
    #[error("An error occurred while converting bytes to UTF-8 for decryption: {0}")]
    FromUtf8(#[from] string::FromUtf8Error),

    /// The key ID in the encrypted data did not match the expected key ID.
    #[error("There was a mismatch in the key id while decrypting data (got: {0}, expected: {1})")]
    MismatchedKeyId(u32, u32),
}

/// Trait for types that can be encrypted.
pub trait Encrypt<T> {
    /// Encrypts this value using the provided encryption key.
    fn encrypt(self, encryption_key: &EncryptionKey) -> Result<T, EncryptionError>;
}

/// Trait for types that can be decrypted.
pub trait Decrypt<T> {
    /// Decrypts this value using the provided encryption key.
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<T, DecryptionError>;
}

/// Encryption key with identifier for key management.
pub struct EncryptionKey {
    /// Unique identifier for the key.
    pub id: u32,
    /// The key material used for encryption and decryption.
    pub key: RandomizedNonceKey,
}

/// Encrypted value with metadata for decryption.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct EncryptedValue {
    /// Identifier of the key used for encryption.
    pub id: u32,
    /// Base64-encoded nonce used during encryption.
    pub nonce: String,
    /// Base64-encoded encrypted value.
    pub value: String,
}

/// Encrypts a string using AES-256-GCM encryption.
///
/// Returns an [`EncryptedValue`] containing the key ID, nonce, and encrypted data,
/// all base64-encoded for safe storage and transmission.
pub fn encrypt_text(
    value: String,
    encryption_key: &EncryptionKey,
) -> Result<EncryptedValue, EncryptionError> {
    let (encrypted_password, nonce) = encrypt(value.as_bytes(), &encryption_key.key)?;
    let encoded_encrypted_password = BASE64_STANDARD.encode(encrypted_password);
    let encoded_nonce = BASE64_STANDARD.encode(nonce.as_ref());

    Ok(EncryptedValue {
        id: encryption_key.id,
        nonce: encoded_nonce,
        value: encoded_encrypted_password,
    })
}

/// Decrypts an [`EncryptedValue`] back to the original string.
///
/// Validates the key ID matches before attempting decryption. Returns the original
/// plaintext string if decryption succeeds.
pub fn decrypt_text(
    encrypted_value: EncryptedValue,
    encryption_key: &EncryptionKey,
) -> Result<String, DecryptionError> {
    if encrypted_value.id != encryption_key.id {
        return Err(DecryptionError::MismatchedKeyId(
            encrypted_value.id,
            encryption_key.id,
        ));
    }

    let encrypted_value_bytes = BASE64_STANDARD.decode(encrypted_value.value)?;
    let nonce = Nonce::try_assume_unique_for_key(&BASE64_STANDARD.decode(encrypted_value.nonce)?)?;

    let decrypted_value_bytes = decrypt(encrypted_value_bytes, nonce, &encryption_key.key)?;

    let decrypted_value = String::from_utf8(decrypted_value_bytes)?;

    Ok(decrypted_value)
}

/// Encrypts bytes using AES-256-GCM with a randomized nonce.
///
/// Returns the encrypted data and the nonce used for encryption.
fn encrypt(
    plaintext: &[u8],
    key: &RandomizedNonceKey,
) -> Result<(Vec<u8>, Nonce), aws_lc_rs::error::Unspecified> {
    let mut in_out = plaintext.to_vec();
    let nonce = key.seal_in_place_append_tag(Aad::empty(), &mut in_out)?;

    Ok((in_out, nonce))
}

/// Decrypts AES-256-GCM encrypted data using the key and nonce.
///
/// Returns the original plaintext bytes.
fn decrypt(
    mut ciphertext: Vec<u8>,
    nonce: Nonce,
    key: &RandomizedNonceKey,
) -> Result<Vec<u8>, aws_lc_rs::error::Unspecified> {
    let plaintext = key.open_in_place(nonce, Aad::empty(), &mut ciphertext)?;

    Ok(plaintext.to_vec())
}

/// Generates a cryptographically secure random encryption key.
///
/// Creates a new [`RandomizedNonceKey`] for AES-256-GCM encryption using
/// secure random bytes.
///
/// # Panics
/// Panics if `T` doesn't match the required key length for AES-256-GCM.
pub fn generate_random_key<const T: usize>()
-> Result<RandomizedNonceKey, aws_lc_rs::error::Unspecified> {
    let mut key_bytes = [0u8; T];
    fill(&mut key_bytes)?;

    let key = RandomizedNonceKey::new(&AES_256_GCM, &key_bytes)?;

    Ok(key)
}
