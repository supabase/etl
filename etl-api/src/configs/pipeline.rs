use etl_config::shared::{BatchConfig, PgConnectionConfig, PipelineConfig};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

const DEFAULT_BATCH_MAX_SIZE: usize = 1000000;
const DEFAULT_BATCH_MAX_FILL_MS: u64 = 10000;
const DEFAULT_TABLE_ERROR_RETRY_DELAY_MS: u64 = 10000;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct FullApiPipelineConfig {
    #[schema(example = "my_publication")]
    pub publication_name: String,
    pub batch: Option<BatchConfig>,
    #[schema(example = 1000)]
    pub table_error_retry_delay_ms: Option<u64>,
    #[schema(example = 4)]
    pub max_table_sync_workers: Option<u16>,
}

impl From<StoredPipelineConfig> for FullApiPipelineConfig {
    fn from(value: StoredPipelineConfig) -> Self {
        Self {
            publication_name: value.publication_name,
            batch: Some(value.batch),
            table_error_retry_delay_ms: Some(value.table_error_retry_delay_ms),
            max_table_sync_workers: Some(value.max_table_sync_workers),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct PartialApiPipelineConfig {
    #[schema(example = "my_publication")]
    pub publication_name: Option<String>,
    #[schema(example = r#"{"max_size": 1000000, "max_fill_ms": 10000}"#)]
    pub batch: Option<BatchConfig>,
    #[schema(example = 1000)]
    pub table_error_retry_delay_ms: Option<u64>,
    #[schema(example = 4)]
    pub max_table_sync_workers: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StoredPipelineConfig {
    pub publication_name: String,
    pub batch: BatchConfig,
    pub table_error_retry_delay_ms: u64,
    pub max_table_sync_workers: u16,
}

impl StoredPipelineConfig {
    pub fn into_etl_config(
        self,
        pipeline_id: u64,
        pg_connection_config: PgConnectionConfig,
    ) -> PipelineConfig {
        PipelineConfig {
            id: pipeline_id,
            publication_name: self.publication_name,
            pg_connection: pg_connection_config,
            batch: self.batch,
            table_error_retry_delay_ms: self.table_error_retry_delay_ms,
            max_table_sync_workers: self.max_table_sync_workers,
        }
    }

    pub fn merge(&mut self, partial: PartialApiPipelineConfig) {
        if let Some(value) = partial.publication_name {
            self.publication_name = value;
        }

        if let Some(value) = partial.batch {
            self.batch = value;
        }

        if let Some(value) = partial.table_error_retry_delay_ms {
            self.table_error_retry_delay_ms = value;
        }

        if let Some(value) = partial.max_table_sync_workers {
            self.max_table_sync_workers = value;
        }
    }
}

impl From<FullApiPipelineConfig> for StoredPipelineConfig {
    fn from(value: FullApiPipelineConfig) -> Self {
        Self {
            publication_name: value.publication_name,
            batch: value.batch.unwrap_or(BatchConfig {
                max_size: DEFAULT_BATCH_MAX_SIZE,
                max_fill_ms: DEFAULT_BATCH_MAX_FILL_MS,
            }),
            table_error_retry_delay_ms: value
                .table_error_retry_delay_ms
                .unwrap_or(DEFAULT_TABLE_ERROR_RETRY_DELAY_MS),
            max_table_sync_workers: value.max_table_sync_workers.unwrap_or_default(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use etl_config::shared::BatchConfig;

    #[test]
    fn test_stored_pipeline_config_serialization() {
        let config = StoredPipelineConfig {
            publication_name: "test_publication".to_string(),
            batch: BatchConfig {
                max_size: 1000,
                max_fill_ms: 5000,
            },
            table_error_retry_delay_ms: 2000,
            max_table_sync_workers: 4,
        };

        let json = serde_json::to_string(&config).unwrap();
        let deserialized: StoredPipelineConfig = serde_json::from_str(&json).unwrap();

        assert_eq!(config.publication_name, deserialized.publication_name);
        assert_eq!(config.batch.max_size, deserialized.batch.max_size);
        assert_eq!(
            config.table_error_retry_delay_ms,
            deserialized.table_error_retry_delay_ms
        );
        assert_eq!(
            config.max_table_sync_workers,
            deserialized.max_table_sync_workers
        );
    }

    #[test]
    fn test_full_api_pipeline_config_conversion() {
        let full_config = FullApiPipelineConfig {
            publication_name: "test_publication".to_string(),
            batch: None,
            table_error_retry_delay_ms: None,
            max_table_sync_workers: None,
        };

        let stored: StoredPipelineConfig = full_config.clone().into();
        let back_to_full: FullApiPipelineConfig = stored.into();

        assert_eq!(full_config.publication_name, back_to_full.publication_name);
    }

    #[test]
    fn test_full_api_pipeline_config_defaults() {
        let full_config = FullApiPipelineConfig {
            publication_name: "test_publication".to_string(),
            batch: None,
            table_error_retry_delay_ms: None,
            max_table_sync_workers: None,
        };

        let stored: StoredPipelineConfig = full_config.into();

        assert_eq!(stored.batch.max_size, DEFAULT_BATCH_MAX_SIZE);
        assert_eq!(stored.batch.max_fill_ms, DEFAULT_BATCH_MAX_FILL_MS);
        assert_eq!(
            stored.table_error_retry_delay_ms,
            DEFAULT_TABLE_ERROR_RETRY_DELAY_MS
        );
        assert_eq!(stored.max_table_sync_workers, 0);
    }

    #[test]
    fn test_partial_api_pipeline_config_merge() {
        let mut stored = StoredPipelineConfig {
            publication_name: "old_publication".to_string(),
            batch: BatchConfig {
                max_size: 500,
                max_fill_ms: 2000,
            },
            table_error_retry_delay_ms: 1000,
            max_table_sync_workers: 2,
        };

        let partial = PartialApiPipelineConfig {
            publication_name: Some("new_publication".to_string()),
            batch: Some(BatchConfig {
                max_size: 2000,
                max_fill_ms: 8000,
            }),
            table_error_retry_delay_ms: Some(5000),
            max_table_sync_workers: None,
        };

        stored.merge(partial);

        assert_eq!(stored.publication_name, "new_publication");
        assert_eq!(stored.batch.max_size, 2000);
        assert_eq!(stored.batch.max_fill_ms, 8000);
        assert_eq!(stored.table_error_retry_delay_ms, 5000);
        assert_eq!(stored.max_table_sync_workers, 2);
    }
}
