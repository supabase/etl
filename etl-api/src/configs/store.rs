use serde::Serialize;
use serde::de::DeserializeOwned;

/// Market trait that has to be implemented by configs that can be stored in the database.
///
/// With this trait we can enforce at compile time which structs can actually be stored and avoid
/// storing the wrong struct.
pub trait Store: Serialize + DeserializeOwned {}
