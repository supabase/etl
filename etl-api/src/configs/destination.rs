use etl_config::SerializableSecretString;
use etl_config::shared::DestinationConfig;
use secrecy::ExposeSecret;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::configs::encryption::{
    Decrypt, DecryptionError, Encrypt, EncryptedValue, EncryptionError, EncryptionKey,
    decrypt_text, encrypt_text,
};
use crate::configs::store::Store;

const DEFAULT_MAX_CONCURRENT_STREAMS: usize = 8;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum FullApiDestinationConfig {
    Memory,
    BigQuery {
        #[schema(example = "my-gcp-project")]
        project_id: String,
        #[schema(example = "my_dataset")]
        dataset_id: String,
        #[schema(example = "{\"type\": \"service_account\", \"project_id\": \"my-project\"}")]
        service_account_key: SerializableSecretString,
        #[schema(example = 15)]
        max_staleness_mins: Option<u16>,
        #[schema(example = 8)]
        max_concurrent_streams: Option<usize>,
    },
}

impl From<StoredDestinationConfig> for FullApiDestinationConfig {
    fn from(value: StoredDestinationConfig) -> Self {
        match value {
            StoredDestinationConfig::Memory => Self::Memory,
            StoredDestinationConfig::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams: Some(max_concurrent_streams),
            },
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StoredDestinationConfig {
    Memory,
    BigQuery {
        project_id: String,
        dataset_id: String,
        service_account_key: SerializableSecretString,
        max_staleness_mins: Option<u16>,
        max_concurrent_streams: usize,
    },
}

impl StoredDestinationConfig {
    pub fn into_etl_config(self) -> DestinationConfig {
        match self {
            Self::Memory => DestinationConfig::Memory,
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => DestinationConfig::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            },
        }
    }
}

impl From<FullApiDestinationConfig> for StoredDestinationConfig {
    fn from(value: FullApiDestinationConfig) -> Self {
        match value {
            FullApiDestinationConfig::Memory => Self::Memory,
            FullApiDestinationConfig::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams: max_concurrent_streams
                    .unwrap_or(DEFAULT_MAX_CONCURRENT_STREAMS),
            },
        }
    }
}

impl Encrypt<EncryptedStoredDestinationConfig> for StoredDestinationConfig {
    fn encrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<EncryptedStoredDestinationConfig, EncryptionError> {
        match self {
            Self::Memory => Ok(EncryptedStoredDestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => {
                let encrypted_service_account_key = encrypt_text(
                    service_account_key.expose_secret().to_owned(),
                    encryption_key,
                )?;

                Ok(EncryptedStoredDestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key: encrypted_service_account_key,
                    max_staleness_mins,
                    max_concurrent_streams,
                })
            }
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EncryptedStoredDestinationConfig {
    Memory,
    BigQuery {
        project_id: String,
        dataset_id: String,
        service_account_key: EncryptedValue,
        max_staleness_mins: Option<u16>,
        max_concurrent_streams: usize,
    },
}

impl Store for EncryptedStoredDestinationConfig {}

impl Decrypt<StoredDestinationConfig> for EncryptedStoredDestinationConfig {
    fn decrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<StoredDestinationConfig, DecryptionError> {
        match self {
            Self::Memory => Ok(StoredDestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key: encrypted_service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => {
                let service_account_key = SerializableSecretString::from(decrypt_text(
                    encrypted_service_account_key,
                    encryption_key,
                )?);

                Ok(StoredDestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key,
                    max_staleness_mins,
                    max_concurrent_streams,
                })
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configs::encryption::{EncryptionKey, generate_random_key};

    #[test]
    fn test_stored_destination_config_serialization() {
        let config = StoredDestinationConfig::BigQuery {
            project_id: "test-project".to_string(),
            dataset_id: "test_dataset".to_string(),
            service_account_key: SerializableSecretString::from("{\"test\": \"key\"}".to_string()),
            max_staleness_mins: Some(15),
            max_concurrent_streams: 8,
        };

        let json = serde_json::to_string(&config).unwrap();
        let deserialized: StoredDestinationConfig = serde_json::from_str(&json).unwrap();

        match (config, deserialized) {
            (
                StoredDestinationConfig::BigQuery { project_id: p1, .. },
                StoredDestinationConfig::BigQuery { project_id: p2, .. },
            ) => {
                assert_eq!(p1, p2);
            }
            _ => panic!("Config types don't match"),
        }
    }

    #[test]
    fn test_stored_destination_config_encryption_decryption() {
        let config = StoredDestinationConfig::BigQuery {
            project_id: "test-project".to_string(),
            dataset_id: "test_dataset".to_string(),
            service_account_key: SerializableSecretString::from("{\"test\": \"key\"}".to_string()),
            max_staleness_mins: Some(15),
            max_concurrent_streams: 8,
        };

        let key = EncryptionKey {
            id: 1,
            key: generate_random_key::<32>().unwrap(),
        };

        let encrypted = config.clone().encrypt(&key).unwrap();
        let decrypted = encrypted.decrypt(&key).unwrap();

        match (config, decrypted) {
            (
                StoredDestinationConfig::BigQuery {
                    project_id: p1,
                    dataset_id: d1,
                    service_account_key: key1,
                    max_staleness_mins: staleness1,
                    max_concurrent_streams: streams1,
                },
                StoredDestinationConfig::BigQuery {
                    project_id: p2,
                    dataset_id: d2,
                    service_account_key: key2,
                    max_staleness_mins: staleness2,
                    max_concurrent_streams: streams2,
                },
            ) => {
                assert_eq!(p1, p2);
                assert_eq!(d1, d2);
                assert_eq!(staleness1, staleness2);
                assert_eq!(streams1, streams2);
                // Assert that service account key was encrypted and decrypted correctly
                assert_eq!(key1.expose_secret(), key2.expose_secret());
            }
            _ => panic!("Config types don't match"),
        }
    }

    #[test]
    fn test_full_api_destination_config_conversion() {
        let full_config = FullApiDestinationConfig::BigQuery {
            project_id: "test-project".to_string(),
            dataset_id: "test_dataset".to_string(),
            service_account_key: SerializableSecretString::from("{\"test\": \"key\"}".to_string()),
            max_staleness_mins: Some(15),
            max_concurrent_streams: None,
        };

        let stored: StoredDestinationConfig = full_config.clone().into();
        let back_to_full: FullApiDestinationConfig = stored.into();

        match (full_config, back_to_full) {
            (
                FullApiDestinationConfig::BigQuery { project_id: p1, .. },
                FullApiDestinationConfig::BigQuery { project_id: p2, .. },
            ) => {
                assert_eq!(p1, p2);
            }
            _ => panic!("Config types don't match"),
        }
    }
}
