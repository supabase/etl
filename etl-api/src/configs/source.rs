use etl_config::SerializableSecretString;
use etl_config::shared::{PgConnectionConfig, TlsConfig};
use secrecy::ExposeSecret;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::configs::encryption::{
    Decrypt, DecryptionError, Encrypt, EncryptedValue, EncryptionError, EncryptionKey,
    decrypt_text, encrypt_text,
};

const DEFAULT_TLS_TRUSTED_ROOT_CERTS: &str = "";
const DEFAULT_TLS_ENABLED: bool = false;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub struct FullApiSourceConfig {
    #[schema(example = "localhost")]
    pub host: String,
    #[schema(example = 5432)]
    pub port: u16,
    #[schema(example = "mydb")]
    pub name: String,
    #[schema(example = "postgres")]
    pub username: String,
    #[schema(example = "secret123")]
    pub password: Option<SerializableSecretString>,
}

impl From<StoredSourceConfig> for FullApiSourceConfig {
    fn from(stored_source_config: StoredSourceConfig) -> Self {
        Self {
            host: stored_source_config.host,
            port: stored_source_config.port,
            name: stored_source_config.name,
            username: stored_source_config.username,
            password: stored_source_config.password,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub struct StrippedApiSourceConfig {
    #[schema(example = "localhost")]
    pub host: String,
    #[schema(example = 5432)]
    pub port: u16,
    #[schema(example = "mydb")]
    pub name: String,
    #[schema(example = "postgres")]
    pub username: String,
}

impl From<StoredSourceConfig> for StrippedApiSourceConfig {
    fn from(source: StoredSourceConfig) -> Self {
        Self {
            host: source.host,
            port: source.port,
            name: source.name,
            username: source.username,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct StoredSourceConfig {
    pub host: String,
    pub port: u16,
    pub name: String,
    pub username: String,
    pub password: Option<SerializableSecretString>,
}

impl StoredSourceConfig {
    pub fn into_connection_config(self) -> PgConnectionConfig {
        PgConnectionConfig {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: self.password,
            // TODO: enable TLS
            tls: TlsConfig {
                trusted_root_certs: DEFAULT_TLS_TRUSTED_ROOT_CERTS.to_string(),
                enabled: DEFAULT_TLS_ENABLED,
            },
        }
    }

    pub fn into_connection_config_with_tls(self, tls: TlsConfig) -> PgConnectionConfig {
        PgConnectionConfig {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: self.password,
            tls,
        }
    }
}

impl From<FullApiSourceConfig> for StoredSourceConfig {
    fn from(full_api_source_config: FullApiSourceConfig) -> Self {
        Self {
            host: full_api_source_config.host,
            port: full_api_source_config.port,
            name: full_api_source_config.name,
            username: full_api_source_config.username,
            password: full_api_source_config.password,
        }
    }
}

impl Encrypt<EncryptedStoredSourceConfig> for StoredSourceConfig {
    fn encrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<EncryptedStoredSourceConfig, EncryptionError> {
        let mut encrypted_password = None;
        if let Some(password) = self.password {
            encrypted_password = Some(encrypt_text(
                password.expose_secret().to_owned(),
                encryption_key,
            )?);
        }

        Ok(EncryptedStoredSourceConfig {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: encrypted_password,
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedStoredSourceConfig {
    host: String,
    port: u16,
    name: String,
    username: String,
    password: Option<EncryptedValue>,
}

impl Decrypt<StoredSourceConfig> for EncryptedStoredSourceConfig {
    fn decrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<StoredSourceConfig, DecryptionError> {
        let mut decrypted_password = None;
        if let Some(password) = self.password {
            let pwd = decrypt_text(password, encryption_key)?;
            decrypted_password = Some(SerializableSecretString::from(pwd));
        }

        Ok(StoredSourceConfig {
            host: self.host,
            port: self.port,
            name: self.name,
            username: self.username,
            password: decrypted_password,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configs::encryption::{EncryptionKey, generate_random_key};

    #[test]
    fn test_stored_source_config_serialization() {
        let config = StoredSourceConfig {
            host: "localhost".to_string(),
            port: 5432,
            name: "testdb".to_string(),
            username: "user".to_string(),
            password: Some(SerializableSecretString::from("password".to_string())),
        };

        let json = serde_json::to_string(&config).unwrap();
        let deserialized: StoredSourceConfig = serde_json::from_str(&json).unwrap();

        assert_eq!(config.host, deserialized.host);
        assert_eq!(config.port, deserialized.port);
        assert_eq!(config.name, deserialized.name);
        assert_eq!(config.username, deserialized.username);
    }

    #[test]
    fn test_stored_source_config_encryption_decryption() {
        let config = StoredSourceConfig {
            host: "localhost".to_string(),
            port: 5432,
            name: "testdb".to_string(),
            username: "user".to_string(),
            password: Some(SerializableSecretString::from("password".to_string())),
        };

        let key = EncryptionKey {
            id: 1,
            key: generate_random_key::<32>().unwrap(),
        };

        let encrypted = config.clone().encrypt(&key).unwrap();
        let decrypted = encrypted.decrypt(&key).unwrap();

        assert_eq!(config.host, decrypted.host);
        assert_eq!(config.port, decrypted.port);
        assert_eq!(config.name, decrypted.name);
        assert_eq!(config.username, decrypted.username);

        // Assert that password was encrypted and decrypted correctly
        match (config.password, decrypted.password) {
            (Some(original), Some(decrypted_pwd)) => {
                assert_eq!(original.expose_secret(), decrypted_pwd.expose_secret());
            }
            (None, None) => {}
            _ => panic!("Password encryption/decryption failed"),
        }
    }

    #[test]
    fn test_full_api_source_config_conversion() {
        let full_config = FullApiSourceConfig {
            host: "localhost".to_string(),
            port: 5432,
            name: "testdb".to_string(),
            username: "user".to_string(),
            password: Some(SerializableSecretString::from("password".to_string())),
        };

        let stored: StoredSourceConfig = full_config.clone().into();
        let back_to_full: FullApiSourceConfig = stored.into();

        assert_eq!(full_config.host, back_to_full.host);
        assert_eq!(full_config.port, back_to_full.port);
        assert_eq!(full_config.name, back_to_full.name);
        assert_eq!(full_config.username, back_to_full.username);
    }
}
