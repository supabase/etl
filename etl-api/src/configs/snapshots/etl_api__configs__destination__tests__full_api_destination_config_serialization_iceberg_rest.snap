---
source: etl-api/src/configs/destination.rs
expression: full_config
---
{
  "iceberg": {
    "rest": {
      "catalog_uri": "https://catalog.example.com/iceberg",
      "warehouse_name": "my-warehouse",
      "namespace": "my-namespace"
    }
  }
}
